pkg/clusterapi/aws/actuator.go

```diff
@@ -367,35 +340,114 @@ func (a *Actuator) DeleteMachine(machine *clusterv1.Machine) error {
 		return fmt.Errorf("error getting EC2 client: %v", err)
 	}
 
-	terminateInstancesRequest := &ec2.TerminateInstancesInput{
-		InstanceIds: []*string{
-			aws.String(instanceID),
-		},
+	instances, err := a.getRunningInstances(machine, client)
+	if err != nil {
+		return err
 	}
-	terminateInstancesResult, err := client.TerminateInstances(terminateInstancesRequest)
+	if len(instances) == 0 {
+		mLog.Warnf("no instances found to delete for machine")
+		return nil
+	}
+
+	instanceIDs := []*string{}
+	for _, instance := range instances {
+		mLog.WithField("instanceID", *instance.InstanceId).Infof("deleting AWS instance")
+		instanceIDs = append(instanceIDs, instance.InstanceId)
+	}
+
+	err = a.terminateInstances(client, instanceIDs, mLog)
+	return err
+}
+
+// Update attempts to sync machine state with an existing instance. Today this just updates status
+// for details that may have changed. (IPs and hostnames) We do not currently support making any
+// changes to actual machines in AWS. Instead these will be replaced via MachineDeployments.
+func (a *Actuator) Update(cluster *clusterv1.Cluster, machine *clusterv1.Machine) error {
+	mLog := clustoplog.WithMachine(a.logger, machine)
+	mLog.Debugf("updating machine")
+
+	clusterSpec, err := controller.ClusterSpecFromClusterAPI(cluster)
 	if err != nil {
-		// If the instance no longer exists, we need to make sure we don't infinitely
-		// block deletion of the Machine in etcd.
-		if awsErr, ok := err.(awserr.Error); ok {
-			if awsErr.Code() == ec2InstanceIDNotFoundCode {
-				mLog.WithField("instanceID", instanceID).Warnf("AWS instance no longer exists")
-				return nil
+		return err
+	}
+
+	if clusterSpec.Hardware.AWS == nil {
+		return fmt.Errorf("Cluster does not contain an AWS hardware spec")
+	}
+
+	coMachineSetSpec, err := controller.MachineSetSpecFromClusterAPIMachineSpec(&machine.Spec)
+	if err != nil {
+		return err
+	}
+
+	region := clusterSpec.Hardware.AWS.Region
+	mLog.WithField("region", region).Debugf("obtaining EC2 client for region")
+	client, err := a.ec2Client(coMachineSetSpec, machine.Namespace, region)
+	if err != nil {
+		return fmt.Errorf("unable to obtain EC2 client: %v", err)
+	}
+
+	instances, err := a.getRunningInstances(machine, client)
+	if err != nil {
+		return err
+	}
+
+	if len(instances) == 0 {
+		// Parent controller should prevent this from ever happening by calling Exists and then Create.
```

Comment:
This could happen if the instance was deleted by a third party in between the call to Exists and the call to Update.